# Python Programming Assignment for Certification  

## Overview  
This assignment consists of 15 coding exercises designed to test and strengthen my understanding of Python programming. The exercises are divided into three sections:  

- **Section 1: Python Operators and Control Structures** – Covers loops, conditionals, and basic operations.  
- **Section 2: Lists and Dictionaries** – Focuses on working with data structures.  
- **Section 3: Advanced Topics** – Explores recursion, generators, and mathematical operations.  

The goal of this assignment is to enhance my ability to apply Python concepts in practical scenarios. I will write and document solutions using Jupyter Notebook, ensuring clarity and readability.  


### Section 1: Operators and Control Structures.

#### Question 1: Write a Python program to calculate the sum of all numbers in a list using a for loop.
```
Sample Input: 1 2 3 4 5
Sample Output: 15

```

In [17]:
# List of numbers
numbers = list(map(int, input("Enter numbers seperated by spaces: ").split()))

total = 0

for num in numbers:
    total += num
print(total)

15


Explanation:
* Initialize total = 0.
* Loop through numbers and add each to total.
* Print total after the loop.

#### Question 2: What is the output of the following code?
```
Sample Input:  x = 5, y = 2
Sample Output: 25

```

In [18]:
x = int(input("Enter the base number (x): "))
y = int(input("Enter the exponent (y): "))
print(x ** y)

25


Why?
* ** is the exponentiation operator in Python.
* x ** y means x raised to the power of y (e.g., 5 ** 2 = 25).

#### Question 3: Write a Python program to check if a number is a palindrome using a while loop.
```
Sample Input: 12321
Sample Output: True

```

A palindrome is a number that remains the same when reversed.
For example:

12321 → Palindrome 
12345 → Not a Palindrome 

In [19]:
def is_palindrome(num):
    original = num
    reversed_num = 0
    temp = num
    
    while temp > 0:
        digit = temp % 10
        reversed_num = reversed_num * 10 + digit
        temp //= 10
        
    return original == reversed_num 
num = int(input("Enter a number: "))
print(is_palindrome(num))

True


Explanation:
* Use % to extract digits one by one.
* Build the reversed number by multiplying the current result by 10 and adding the digit.
* Compare the original and reversed numbers.
* If they match, it's a palindrome; otherwise, it's not.

#### Question 4: What is the output of the following code?

<pre>fruits = ['apple', 'banana', 'cherry']
print(fruits[-1])</pre>


In [20]:
fruit = ['apple', 'banana', 'cherry']
print(fruit[-1])

cherry


Why?
* Negative indexing in Python accesses elements from the end of a list.
* fruits[-1] gets the last item.
* In ['apple', 'banana', 'cherry'], fruits[-1] returns 'cherry'.

#### Question 5: Write a Python program to calculate the average of all numbers in a list using a for loop.
```
Sample Input: 1 2 3 4 5
Sample Output: 3.0

```

In [21]:
numbers = list(map(float, input("Enter numbers seperated by spaces: ").split()))
total = 0
for num in numbers:
    total += num
    
average = total / len(numbers)
print(average)

3.0


Explanation:
* Set total = 0 to store the sum.
* Loop through numbers, adding each to total.
* Compute the average as total / len(numbers).
* Print the result.

### Section 2: Lists & Dictionaries.

#### Question 6: Write a Python program to create a dictionary with student names as keys and their grades as values.
```
Sample Input: Students = John, Alice, Bob
              Grades = 90, 85, 95
Sample Output: {'John': 90, ' Alice': 85, ' Bob': 95}

```

In [9]:
students = input("Enter student names seperated by commas: ").split(',')
grades = list(map(int, input("Enter grades seperated by commas: ").split(',')))
student_dict = dict(zip(students, grades))
print(student_dict)

{'John': 90, ' Alice': 85, ' Bob': 95}


Explanation:
* zip(students, grades) pairs each student with their grade.
* dict(zip(...)) converts these pairs into a dictionary.
* The result has student names as keys and grades as values.








#### Question 7: What is the output of the following code?
<pre>numbers = [1, 2, 3, 4, 5]
print(numbers[1:3])</pre>


In [22]:
numbers = [1, 2, 3, 4, 5]
print(numbers[1:3])

[2, 3]


Why?
* List slicing extracts a part of a list.
* numbers[1:3] starts at index 1 (2) and stops before index 3 (4).
* The result is [2, 3].

#### Question 8: Write a Python program to sort a list of numbers in ascending order using the sorted() function.
```
 Sample Input: numbers = [4, 2, 9, 6, 5]
 Sample Output: [2, 4, 5, 6, 9]
```

In [11]:
numbers = list(map(int, input("Enter numbers seperated by spaces: ").split()))
sorted_numbers = sorted(numbers)
print(sorted_numbers)

[2, 4, 5, 6, 9]


Explanation:
* sorted() returns a new sorted list without changing the original.
* sorted(numbers) arranges numbers in ascending order.
* The result is stored in sorted_numbers and printed.

#### Question 9: What is the output of the following code?
<pre>fruits = {'apple': 5, 'banana': 10, 'cherry': 15} 
print(fruits.get('banana'))</pre>


In [23]:
fruits = {'apple' : 5, 'banana': 10, 'cherry' : 15}
print(fruits.get('banana'))

10


Why?
* Dictionaries store key-value pairs in Python.
* .get('banana') fetches the value for 'banana'.
* Since 'banana' exists, it returns 10.

#### Question 10: Write a Python program to create a list of squares of all numbers from 1 to 10 using a list comprehension.
```
Sample Input: Start = 1
              End = 10
Sample Output: Squares:  [1, 4, 9, 16, 25, 36, 49, 64, 81, 100]

```

In [24]:
start = int(input("Enter the start number: "))
end = int(input("Enter the end number: "))
squares = [x ** 2 for x in range(start, end + 1)]
print("Squares: ", squares)

Squares:  [1, 4, 9, 16, 25, 36, 49, 64, 81, 100]


Explanation:
* List comprehension [x**2 for x in range(1, 11)] creates a list of squares.
* range(1, 11) iterates from 1 to 10.
* Each x is squared (x**2) and stored in the list.

### Section 3: Advanced Topics.

#### Question 11: Write a Python program to calculate the factorial of a number using recursion.
```
Sample Input: n = 5
Sample Output: 120
```

In [25]:
def factorial(n):
    if n == 0 or n == 1:
        return 1
    else:
        return n * factorial(n - 1)
    
num = int(input("Enter a number: "))
if num < 0:
    print("Factorial is not defined for negative numbers.")
else:
    print(f"Factorial of {num} is: {factorial(num)}")

Factorial of 5 is: 120


Explanation:
* Base case: If n == 0 or n == 1, return 1 (since 0! = 1! = 1).
* Recursive case: Multiply n by factorial(n - 1) until reaching the base case.
* Example (n = 5):
factorial(5) = 5 * factorial(4)
factorial(4) = 4 * factorial(3)
factorial(3) = 3 * factorial(2)
factorial(2) = 2 * factorial(1)
factorial(1) = 1 (Base case)
The result: 5 * 4 * 3 * 2 * 1 = 120.

#### Question 12: What is the output of the following code?
<pre>numbers = [1, 2, 3, 4, 5]
print(list(reversed(numbers)))
</pre>

In [28]:
numbers = [1, 2, 3, 4, 5]
print(list(reversed(numbers)))

[5, 4, 3, 2, 1]


Why?
* reversed() creates an iterator that goes through the list backward.
* list(reversed(numbers)) converts it into a reversed list.
* [1, 2, 3, 4, 5] becomes [5, 4, 3, 2, 1].

#### Question 13: Write a Python program to create a generator that yields the Fibonacci sequence up to the nth term.
```
Sample Input: n = 10
Sample Output: [0, 1, 1, 2, 3, 5, 8, 13, 21, 34]
```

In [26]:
def fibonacci_generator(n):
    a, b = 0, 1
    for _ in range(n):
        yield a
        a, b = b, a + b
        
n = int(input("Enter the number of Fibonacci terms: "))
if n < 0:
    print("Please enter a non-negative numbers.")
else:
   print(list(fibonacci_generator(n)))

[0, 1, 1, 2, 3, 5, 8, 13, 21, 34]


Explanation of Code Execution:

* yield a returns the current Fibonacci number and pauses execution.
* a, b = b, a + b updates values to the next Fibonacci number.
* The loop runs n times, generating n Fibonacci numbers.

#### Question 14: What is the output of the following code?
<pre>import math
print(math.pi)
</pre>

In [30]:
import math
print(math.pi)

3.141592653589793


Why?
* The math module offers mathematical functions and constants.
* math.pi gives the constant π (≈ 3.14159).
* It's widely used in geometry and trigonometry.

#### Question 15: Write a Python program to print an inverted star pyramid of height 5.
<pre>*********
 *******
  *****
   ***
    *
</pre>

In [31]:
def inverted_pyramid(height):
    for i in range(height, 0, -1):
        spaces = ' ' * (height - i)
        stars = '*' * (2 * i -1)
        print(spaces + stars)
        
height = 5
inverted_pyramid(height)

*********
 *******
  *****
   ***
    *


* Explanation:
* 1.Loop from height down to 1 (`range(height, 0, -1)`)
   - Starts at `height = 5`, decreases to `1`.

* 2.Calculate spaces (`' ' * (height - i)`)
   - Number of spaces increases as `i` decreases.
   - Example:
     ```
     Row 1: 0 spaces
     Row 2: 1 space
     Row 3: 2 spaces
     Row 4: 3 spaces
     Row 5: 4 spaces
     ```

* 3.Calculate stars (`'*' * (2 * i - 1)`)
   - Number of stars follows the pattern:
     ```
     Row 1: 9 stars
     Row 2: 7 stars
     Row 3: 5 stars
     Row 4: 3 stars
     Row 5: 1 star
     ```
```

```
* 4.Print `spaces + stars` to form the pyramid shape.

  Final Output:
    
```
<pre>*********
 *******
  *****
   ***
    *</pre>
```
```